File: Reporting/JavaScript/protractor-sample/mocha/conf.js
const Reporting = require('perfecto-reporting');
const inherits = require('mocha').utils.inherits;
const Base = require('mocha').reporters.Base;

var reportingClient;

const PerfectoMochaReporter = function (runner) {
  Base.call(this, runner);
  var passes = 0;
  var failures = 0;

  runner.on('test', function (test) {
    reportingClient.testStart(test.fullTitle(), ['Mocha' , 'test_integrate_mocha_reporting']);
  });

  runner.on('pass', function (test) {
    reportingClient.testStop({
      status: Reporting.Constants.results.passed
    });
  });

  runner.on('fail', function (test, err) {
    console.log(`${err.message} ${err.stack}`);
    reportingClient.testStop({
      status: Reporting.Constants.results.failed,
      message: err.message
    });
  });
}

// Custopm reporter inherit from Mocha base reporter 
inherits(PerfectoMochaReporter, Base);

exports.config = {
  //Remote address
  seleniumAddress: 'https://MY_HOST.perfectomobile.com/nexperience/perfectomobile/wd/hub',

  //Capabilities to be passed to the webdriver instance.
  capabilities: {
    browserName: 'chrome',
    user: 'MY_USER',
    password: 'MY_PASS',
    platformName: 'Android'
  },

  //Framework to use
  framework: 'mocha',

  mochaOpts: {
    reporter: PerfectoMochaReporter,
    timeout: 120000, // default suite timeout for Mocha is 2000ms. Most Selenium tests take longer
    slow: 3000
  },

  // Spec patterns are relative to the current working directly when
  // protractor is called.
  specs: ['spec.js'],

  onComplete: function () {
    // Output report URL
    return reportingClient.getReportUrl().then(
      function (url) {
        console.log(`Report url = ${url}`);
      }
    );
  },

  onPrepare: function () {
    reportingClient = new Reporting.Perfecto.PerfectoReportingClient(new Reporting.Perfecto.PerfectoExecutionContext({
      webdriver: browser.driver,
      tags: ['javascript driver']
    }));
    browser.reportingClient = reportingClient;
  }
}

File: Reporting/Python/Conf.py
import unittest
from selenium import webdriver
from perfecto import PerfectoExecutionContext,TestResultFactory,TestContext,PerfectoReportiumClient


class TestConf(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        self.user = 'My_User'
        self.password = 'My_Pass'
        self.host = 'My_Host.perfectomobile.com'
        self.driver = None
        self.reporting_client = None

        super(TestConf, self).__init__(*args, **kwargs)

    def setUp(self):
        capabilities = {
            'platformName': 'Android',
            'deviceName': '',
            'user': self.user,
            'password': self.password
        }
        self.driver = webdriver.Remote('https://' + self.host + '/nexperience/perfectomobile/wd/hub', capabilities)
        self.create_reporting_client()
        self.reporting_client.test_start(self.id(),
                                         TestContext('Tag1', 'Tag2', 'Tag3'))

    def run(self, result=None):
        self.currentResult = result  # remember result for use in tearDown
        unittest.TestCase.run(self, result)  # call superclass run method

    def tearDown(self):
        try:
            if self.currentResult.wasSuccessful():
                self.reporting_client.test_stop(TestResultFactory.create_success())
            else:
                self.reporting_client.test_stop(TestResultFactory.create_failure(self.currentResult.errors,
                                                                                 self.currentResult.failures))
            # Print report's url
            print 'Report-Url: ' + self.reporting_client.report_url() + '\n'
        except Exception as e:
            print e.message

        self.driver.quit()

    def create_reporting_client(self):
        perfecto_execution_context = PerfectoExecutionContext(self.driver)
        self.reporting_client = PerfectoReportiumClient(perfecto_execution_context)
